### PR TITLE
[Macros] Create type refinement context for MacroExpansionDecl

### PR DESCRIPTION
**Pull request by rintaro**
rdar://107386648

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/AST/TypeRefinementContext.cpp [hunk 1]
```diff
@@ -53,6 +53,13 @@ TypeRefinementContext::createRoot(SourceFile *SF,
   if (auto parentExpansion = SF->getMacroExpansion()) {
     if (auto parentTRC =
             SF->getEnclosingSourceFile()->getTypeRefinementContext()) {
+      // FIXME: Skip `MacroExpansionDecl` context?
```
Comment (DougGregor) [inline lib/AST/TypeRefinementContext.cpp:56]: I think you don't need this.

**Review by DougGregor — APPROVED** [review]

File: lib/AST/ASTWalker.cpp [hunk 2]
```diff
@@ -1482,7 +1482,7 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
   
   bool shouldSkip(Decl *D) {
     if (!Walker.shouldWalkMacroArgumentsAndExpansion().second &&
-        D->isInMacroExpansionInContext())
+        D->isInMacroExpansionInContext() && !Walker.Parent.isNull())
```
Comment (rintaro) [inline lib/AST/ASTWalker.cpp:1485]: `TypeRefinementContextBuilder` doesn't walk into expansions, but It does `decl->walk(walker)` on the top level decls of the synthesized buffer.

File: test/Macros/macro_attribute_expansiondecl.swift [hunk 3]
```diff
@@ -57,6 +58,36 @@ public struct LocalFuncAndVarMacro: DeclarationMacro {
...
 
+public struct FuncWithClosureMacro: DeclarationMacro {
+  public static func expansion(
+    of node: some FreestandingMacroExpansionSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [DeclSyntax] {
+
+    return ["func LocalFunc() {}", "var LocalVar: Int { 1 }"]
+  }
+}
```
Comment (rintaro) [inline test/Macros/macro_attribute_expansiondecl.swift:69]: Unrelated. Will remove this

File: test/Macros/macro_attribute_expansiondecl.swift [hunk 4]
```diff
@@ -97,3 +128,20 @@ func testLocal() {
...
+
+@freestanding(declaration, names: named(fromClosure)) macro funcFromClosureMacro(_: String, _: () -> Void) = #externalMacro(module: "MacroDefinition", type: "FuncFromClosureMacro")
+
+@available(macOS 99, *)
+func APIFrom99() -> String { "" }
+@available(macOS 999, *)
+func APIFrom999() -> String { "" }
+
+@available(macOS 99, *)
+#funcFromClosureMacro(APIFrom99()) {
```
Comment (rintaro) [inline test/Macros/macro_attribute_expansiondecl.swift:140]: Add test case without this attribute. to check if it's diagnosed properly

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**